beacon-chain/state/v3/setters_misc.go

```diff
@@ -0,0 +1,181 @@
+		return ErrNilInnerState
+	}
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	fk, ok := proto.Clone(val).(*ethpb.Fork)
+	if !ok {
+		return errors.New("proto.Clone did not return a fork proto")
+	}
+	b.state.Fork = fk
+	b.markFieldAsDirty(fork)
+	return nil
+}
+
+// SetHistoricalRoots for the beacon state. Updates the entire
+// list to a new value by overwriting the previous one.
+func (b *BeaconState) SetHistoricalRoots(val [][]byte) error {
+	if !b.hasInnerState() {
+		return ErrNilInnerState
+	}
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.sharedFieldReferences[historicalRoots].MinusRef()
+	b.sharedFieldReferences[historicalRoots] = stateutil.NewRef(1)
+
+	b.state.HistoricalRoots = val
+	b.markFieldAsDirty(historicalRoots)
+	return nil
+}
+
+// AppendHistoricalRoots for the beacon state. Appends the new value
+// to the the end of list.
+func (b *BeaconState) AppendHistoricalRoots(root [32]byte) error {
+	if !b.hasInnerState() {
+		return ErrNilInnerState
+	}
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	roots := b.state.HistoricalRoots
+	if b.sharedFieldReferences[historicalRoots].Refs() > 1 {
+		roots = make([][]byte, len(b.state.HistoricalRoots))
+		copy(roots, b.state.HistoricalRoots)
+		b.sharedFieldReferences[historicalRoots].MinusRef()
+		b.sharedFieldReferences[historicalRoots] = stateutil.NewRef(1)
+	}
+
+	b.state.HistoricalRoots = append(roots, root[:])
+	b.markFieldAsDirty(historicalRoots)
+	return nil
+}
+
+// Recomputes the branch up the index in the Merkle trie representation
+// of the beacon state. This method performs slice reads and the caller MUST
+// hold the lock before calling this method.
+func (b *BeaconState) recomputeRoot(idx int) {
+	hashFunc := hash.CustomSHA256Hasher()
+	layers := b.merkleLayers
+	// The merkle tree structure looks as follows:
+	// [[r1, r2, r3, r4], [parent1, parent2], [root]]
+	// Using information about the index which changed, idx, we recompute
+	// only its branch up the tree.
+	currentIndex := idx
+	root := b.merkleLayers[0][idx]
+	for i := 0; i < len(layers)-1; i++ {
+		isLeft := currentIndex%2 == 0
+		neighborIdx := currentIndex ^ 1
+
+		neighbor := make([]byte, 32)
+		if layers[i] != nil && len(layers[i]) != 0 && neighborIdx < len(layers[i]) {
+			neighbor = layers[i][neighborIdx]
+		}
+		if isLeft {
+			parentHash := hashFunc(append(root, neighbor...))
+			root = parentHash[:]
+		} else {
+			parentHash := hashFunc(append(neighbor, root...))
+			root = parentHash[:]
+		}
+		parentIdx := currentIndex / 2
+		// Update the cached layers at the parent index.
+		layers[i+1][parentIdx] = root
+		currentIndex = parentIdx
+	}
+	b.merkleLayers = layers
+}
+
+func (b *BeaconState) markFieldAsDirty(field stateTypes.FieldIndex) {
+	b.dirtyFields[field] = true
+}
+
+// addDirtyIndices adds the relevant dirty field indices, so that they
+// can be recomputed.
+func (b *BeaconState) addDirtyIndices(index stateTypes.FieldIndex, indices []uint64) {
+	if b.rebuildTrie[index] {
+		return
+	}
```

Comment:
```suggestion
       }
       // Exit early if balance trie computation isn't enabled.
       if !features.Get().EnableBalanceTrieComputation && index == balances {
               return
       }
```
This was added in develop to v2.